sdk/database/helper/credsutil/sql_test.go

```diff
@@ -0,0 +1,97 @@
+package credsutil
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/hashicorp/vault/sdk/database/dbplugin"
+)
+
+func TestGenerateUsername(t *testing.T) {
+	// The length of random alphanumeric string is hardcoded
+	// https://github.com/hashicorp/vault/blob/73b1eed7c9066b121f6d4c77aa2d8ace21b48fef/sdk/database/helper/credsutil/sql.go#L58
+	RandomAlphaNumericLen := 20
+	UnixTimestampLen := len(fmt.Sprint(time.Now().Unix()))
+
+	credsProducer := &SQLCredentialsProducer{
+		DisplayNameLen: 30,
+		RoleNameLen:    8,
+		UsernameLen:    55,
+		Separator:      "-",
+	}
+
+	type testCase struct {
+		displayName        string
+		roleName           string
+		expectedUserPrefix string
+	}
+
+	tests := map[string]testCase{
+		"short display name and role name are fully shown": {
+			displayName: "short",
+			roleName: "ro",
+			expectedUserPrefix: "v-short-ro-",
+		},
+		"long display name is truncated": {
+			displayName: "longlonglonglonglonglonglonglong",
+			roleName: "ro",
+			expectedUserPrefix: "v-longlonglonglonglonglonglonglo-ro-",
+		},
+		"long role name is truncated": {
+			displayName: "short",
+			roleName: "readwrite",
+			expectedUserPrefix: "v-short-readwrit-",
+		},
+		"long display name and role name are truncated": {
+			displayName: "longlonglonglonglonglonglonglong",
+			roleName: "readwrite",
+			expectedUserPrefix: "v-longlonglonglonglonglonglonglo-readwrit-",
+		},
+		"empty display name": {
+			displayName: "",
+			roleName: "ro",
+			expectedUserPrefix: "v-ro-",
+		},
+		"empty role name": {
+			displayName: "short",
+			roleName: "",
+			expectedUserPrefix: "v-short-",
+		},
+		"both display name and role name are empty": {
+			displayName: "",
+			roleName: "",
+			expectedUserPrefix: "v-",
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			usernameConfig := dbplugin.UsernameConfig{
+				DisplayName: test.displayName,
+				RoleName:    test.roleName,
+			}
+
+			username, err := credsProducer.GenerateUsername(usernameConfig)
+			if err != nil {
+				t.Fatalf("err: %s", err)
+			}
+
+			if !strings.HasPrefix(username, test.expectedUserPrefix) {
+				t.Fatalf("Expected username to begin with %s. But got %s", test.expectedUserPrefix, username)
+			}
+
+			var expectedLen int
+			if len(test.expectedUserPrefix) + RandomAlphaNumericLen + 1 + UnixTimestampLen < credsProducer.UsernameLen {
+				expectedLen = len(test.expectedUserPrefix) + RandomAlphaNumericLen + 1 + UnixTimestampLen
+			} else {
+				expectedLen = credsProducer.UsernameLen
+			}
+
+			if len(username) != expectedLen {
```

Comment:
If I'm reading [the code](https://github.com/hashicorp/vault/blob/master/sdk/database/helper/credsutil/sql.go#L41) correctly, `credsProducer.UsernameLen` is the maximum length for a username, so I think you can remove `expectedLen` and all of its associated logic and just replace this if statement with:
```go
if len(username) > credsProducer.UsernameLen {
```
and adjust the error message accordingly. I don't think it's necessarily important to know the exact username length for this test, and including too many internal details about the username generation algorithm will make the test more brittle.

Reply:
@raskchanky Really good suggestion! I was going too far and detailed in testing username length.